runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+	}
+
+	return &conn{
+		Conn: c,
+		db:   d,
+	}, release, nil
+}
+
+func (d *db) CreateTableAsSelect(ctx context.Context, name, query string, opts *CreateTableOptions) error {
+	if opts == nil {
+		opts = &CreateTableOptions{}
+	}
+	d.logger.Debug("create table", slog.String("name", name), slog.Bool("view", opts.View))
+	d.writeMu.Lock()
+	defer d.writeMu.Unlock()
+	conn, release, err := d.acquireWriteConn(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = release()
+	}()
+	return d.createTableAsSelect(ctx, conn, release, name, query, opts)
+}
+
+func (d *db) createTableAsSelect(ctx context.Context, conn *sqlx.Conn, releaseConn func() error, name, query string, opts *CreateTableOptions) error {
+	// check if some older version exists
+	oldVersion, oldVersionExists, _ := tableVersion(d.writePath, name)
+	d.logger.Debug("old version", slog.String("version", oldVersion), slog.Bool("exists", oldVersionExists))
+
+	// create new version directory
+	newVersion := newVersion()
+	newVersionDir := filepath.Join(d.writePath, name, newVersion)
+	err := os.MkdirAll(newVersionDir, fs.ModePerm)
+	if err != nil {
+		return fmt.Errorf("create: unable to create dir %q: %w", name, err)
+	}
+
+	var m meta
+	if opts.View {
+		// create view - validates that SQL is correct
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("CREATE OR REPLACE VIEW %s AS (%s\n)", safeSQLName(name), query))
+		if err != nil {
+			return err
+		}
+
+		m = meta{ViewSQL: query}
+	} else {
+		// create db file
+		dbFile := filepath.Join(newVersionDir, "data.db")
+		safeDBName := safeSQLName(dbName(name))
+
+		// detach existing db
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("DETACH DATABASE IF EXISTS %s", safeDBName), nil)
+		if err != nil {
+			_ = os.RemoveAll(newVersionDir)
+			return fmt.Errorf("create: detach %q db failed: %w", safeDBName, err)
+		}
+
+		// attach new db
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("ATTACH %s AS %s", safeSQLString(dbFile), safeDBName), nil)
+		if err != nil {
+			_ = os.RemoveAll(newVersionDir)
+			return fmt.Errorf("create: attach %q db failed: %w", dbFile, err)
+		}
+
+		// ingest data
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("CREATE OR REPLACE TABLE %s.default AS (%s\n)", safeDBName, query), nil)
+		if err != nil {
+			_ = os.RemoveAll(newVersionDir)
+			return fmt.Errorf("create: create %q.default table failed: %w", safeDBName, err)
+		}
+
+		m = meta{Format: BackupFormatDB}
+	}
+
+	d.writeDirty = true
+	// write meta
+	err = writeMeta(newVersionDir, m)
+	if err != nil {
+		_ = os.RemoveAll(newVersionDir)
+		return err
+	}
+
+	// update version.txt
+	err = os.WriteFile(filepath.Join(d.writePath, name, "version.txt"), []byte(newVersion), fs.ModePerm)
+	if err != nil {
+		_ = os.RemoveAll(newVersionDir)
+		return fmt.Errorf("create: write version file failed: %w", err)
+	}
+
+	// close write handle before syncing read so that temp files or wal files if any are removed
+	err = releaseConn()
+	if err != nil {
+		return err
+	}
```

Comment:
1. It is already called by caller function so no need to call it on multiple error paths. It is also idempotent. `conn.Close` returns error if connection is already closed which we ignore and db.Close is idempotent. It is necessary to call close before syncing so that we don't have to sync `wal` or `temp` file as indicated in the comment (This makes me realise that this is not implemented correctly in `conn.go`).
2. To call `_ = os.RemoveAll(newVersionDir)` the caller will need to get the `newversion` and pass it to the function. 
3. Yeah makes sense. Actually there is a function to read version but not to set version.